packages/medusa/src/services/line-item.ts

```diff
@@ -178,115 +185,205 @@ class LineItemService extends TransactionBaseService {
+   * @param quantity
+   * @param context
+   */
+  async generate<
+    T = string | GenerateInputData | GenerateInputData[],
+    TResult = T extends string
+      ? LineItem
+      : T extends LineItem
+      ? LineItem
+      : LineItem[]
+  >(
+    variantIdOrData: string | T,
+    regionIdOrContext: T extends string ? string : GenerateLineItemContext,
+    quantity?: number,
+    context: GenerateLineItemContext = {}
+  ): Promise<TResult> {
     return await this.atomicPhase_(
       async (transactionManager: EntityManager) => {
-        const [variant, region] = await Promise.all([
-          this.productVariantService_
-            .withTransaction(transactionManager)
-            .retrieve(variantId, {
-              relations: ["product"],
-            }),
-          this.regionService_
-            .withTransaction(transactionManager)
-            .retrieve(regionId),
-        ])
-
-        let unit_price = Number(context.unit_price) < 0 ? 0 : context.unit_price
-
-        let unitPriceIncludesTax = false
-
-        let shouldMerge = false
-
-        if (context.unit_price === undefined || context.unit_price === null) {
-          shouldMerge = true
-          const variantPricing = await this.pricingService_
-            .withTransaction(transactionManager)
-            .getProductVariantPricingById(variant.id, {
-              region_id: region.id,
-              quantity: quantity,
-              customer_id: context?.customer_id,
-              include_discount_prices: true,
-            })
-
-          unitPriceIncludesTax = !!variantPricing.calculated_price_includes_tax
-
-          unit_price = variantPricing.calculated_price ?? undefined
-        }
+        this.validateGenerateArguments(
+          variantIdOrData,
+          regionIdOrContext,
+          quantity
+        )
 
-        const rawLineItem: Partial<LineItem> = {
-          unit_price: unit_price,
-          title: variant.product.title,
-          description: variant.title,
-          thumbnail: variant.product.thumbnail,
-          variant_id: variant.id,
-          quantity: quantity || 1,
-          allow_discounts: variant.product.discountable,
-          is_giftcard: variant.product.is_giftcard,
-          metadata: context?.metadata || {},
-          should_merge: shouldMerge,
-        }
+        const data = isString(variantIdOrData)
+          ? {
+              variantId: variantIdOrData,
+              quantity: quantity as number,
+            }
+          : variantIdOrData
+        const resolvedContext = isString(variantIdOrData)
+          ? context
+          : (regionIdOrContext as GenerateLineItemContext)
+        const regionId = (
+          isString(variantIdOrData)
+            ? regionIdOrContext
+            : resolvedContext.region_id
+        ) as string
+
+        const isDataAnArray = Array.isArray(data)
+        const resolvedData = (
+          isDataAnArray ? data : [data]
+        ) as GenerateInputData[]
```

Comment:
suggestion(non-blocking):
```suggestion
        const resolvedData = (
          Array.isArray(data) ? data : [data]
        ) as GenerateInputData[]
```